### PR TITLE
[BACKLOG-5541] -- Updated api.operation.render method to accept both …

### PR DESCRIPTION
**Pull request by krivera-pentaho**
…a synchronous and asynchronous method for getParameterXml

### CODE REVIEW

**Comment by krivera-pentaho** [comment]
@diogofscmariano @pamval @annapopova1 please review

**Comment by annapopova1** [comment]
The first implementation provided only the synchronous way. Therefore IMO better use the synchronous way by default. It also better for users who already use this API. What do you think guys?

**Comment by diogofscmariano** [comment]
The current implementation is not an issue. The API is not public and we can follow an approach that rewrites it if needed. That is the reason why the API was not published on 6.0.

More global question here... if we need to provide a async and sync way of rendering a prompt why we don't think this from the opposing way? For instance, no matter of what is the way getParameterXml is called (either async or sync), we must provide a much cleaner way for the users to do it. I don't like exposing that info to the user in the method. Why we don't go with something like config the call like jQuery does, via a config on the api or using a json? Also, i don't see the user switching from sync to async. About default or not, IMO we should have the default as async. @pamval @annapopova1 @plagoa @krivera-pentaho  what you guys think?

**Comment by pamval** [comment]
Why don't we just assume the getParameterXML always receives a callback and leave the sync/async definition to whoever implements that function ?

That way, you don't need a boolean on the signature nor a config.

**Comment by krivera-pentaho** [comment]
Sounds good. I will make the changes so that it always receives a callback. If a value is returned, then it is assumed synchronous, otherwise it is asynchronous

**Comment by krivera-pentaho** [comment]
@diogofscmariano @annapopova1 @plagoa @pamval, i updated the PR. please review

File: package-res/resources/web/prompting/api/OperationAPI.js [hunk 1]
```diff
@@ -48,39 +48,62 @@ define(['common-ui/prompting/PromptPanel', 'common-ui/prompting/parameters/Param
      *
      * @name OperationAPI#render
      * @method
-     * @param {String} id HTML object id where the render the panel
-     * @param {Function} getParameterXml The function called when the prompt panel needs to get the new parameter xml
+     * @param {String}    id HTML object id where the render the panel
+     * @param {Function}  getParameterXml The function called when the prompt panel needs to get the new parameter xml.
+     *                                    Receives a callback function. If synchronous, simply return xml. If asynchronous,
+     *                                    call the callback function provided and and pass xml to that.
```
Comment (annapopova1) [inline package-res/resources/web/prompting/api/OperationAPI.js]: double "and"

File: package-res/resources/web/prompting/api/OperationAPI.js [hunk 2]
```diff
@@ -48,39 +48,62 @@ define(['common-ui/prompting/PromptPanel', 'common-ui/prompting/parameters/Param
...
-      var xml = getParameterXml(api);
-      if (!xml) {
-        api.log.error(this._msgs.NO_PARAM_XML, true);
-      }
-
-      var paramDefn = this._parameterParser.parseParameterXml(xml);
-      this._promptPanel = new PromptPanel(id, paramDefn);
+      var getParameterXmlAsync = function(callback) {
+        var xml = getParameterXml(api, callback);
+        if (typeof xml != "undefined") {
```
Comment (annapopova1) [inline package-res/resources/web/prompting/api/OperationAPI.js]: probably unnecessary check because "xml" is checked into callback (row 82).

Comment (krivera-pentaho) [inline package-res/resources/web/prompting/api/OperationAPI.js]: I check specifically for "undefined" here as an indication of whether we are trying to have a synchronous or asynchronous function call. if (!xml) will be for null, or other falsy values.

**Comment by krivera-pentaho** [comment]
@diogofscmariano @pamval @annapopova1 @plagoa please review. I removed the 2nd "and"

File: package-res/resources/web/prompting/api/OperationAPI.js (hunk 2 above)
Comment (annapopova1) [inline package-res/resources/web/prompting/api/OperationAPI.js]: @krivera-pentaho thank you, I understood. I have not more comments.

**Comment by pamval** [comment]
So, what I meant before is that there should be no distinction (from inside the Prompting API) if the handler is synchronous or asynchronous.

That means that we always expect the implementer of the retrieveParameters function to call the callback after the parameter XML has been retrieved. Regardless of whether that retrieval was sync or async.

I'm not sure if I'm using the correct highly technical terms here, @krivera-pentaho , so be patient.

**Comment by krivera-pentaho** [comment]
LOL!!!!! It was too highly technical for me to understand ;)

I understand. So you don't even want the function to even be able to return anything. Always call the callback function. They either call it synchronously or asynchronously, within their implementation for getParameterXml. I will change the implementation.

**Comment by krivera-pentaho** [comment]
@diogofscmariano @annapopova1 @pamval @plagoa please review

**Pull request merged**